pppd/crypto.c

```diff
@@ -112,33 +126,71 @@ void PPP_CIPHER_CTX_free(PPP_CIPHER_CTX *ctx)
 
 int PPP_CipherInit(PPP_CIPHER_CTX *ctx, const PPP_CIPHER *cipher, const unsigned char *key, const unsigned char *iv, int encr)
 {
+    int ret = 0;
     if (ctx && cipher) {
+        ret = 1;
         ctx->is_encr = encr;
         ctx->cipher = *cipher;
         if (ctx->cipher.init_fn) {
-            ctx->cipher.init_fn(ctx, key, iv);
+            ret = ctx->cipher.init_fn(ctx, key, iv);
+            if (!ret) {
+                PPP_crypto_error("Could not initialize cipher");
+            }
         }
-        return 1;
     }
-    return 0;
+    return ret;
 }
 
 int PPP_CipherUpdate(PPP_CIPHER_CTX *ctx, unsigned char *out, int *outl, const unsigned char *in, int inl)
 {
+    int ret = 0;
     if (ctx && ctx->cipher.update_fn) {
-        return ctx->cipher.update_fn(ctx, out, outl, in, inl);
+        ret = ctx->cipher.update_fn(ctx, out, outl, in, inl);
+        if (!ret) {
+            PPP_crypto_error("Could not perform crypto operation");
+        }
     }
-    return 0;
+    return ret;
 }
 
 int PPP_CipherFinal(PPP_CIPHER_CTX *ctx, unsigned char *out, int *outl)
 {
+    int ret = 0;
     if (ctx && ctx->cipher.final_fn) {
-        return ctx->cipher.final_fn(ctx, out, outl);
+        ret = ctx->cipher.final_fn(ctx, out, outl);
+        if (!ret) {
+            PPP_crypto_error("Could not perform final crypto operation");
+        }
+    }
+    return ret;
+}
+
+void
+PPP_crypto_error(char *fmt, ...)
+{
+    va_list args;
+    char buf[1024];
+    int len = sizeof(buf);
+    int off = 0;
+    int err = 0;
+
+    va_start(args, fmt);
+    off = vsnprintf(buf, len, fmt, args);
+    va_end(args);
+
+#ifdef PPP_WITH_OPENSSL
+    err = ERR_peek_error();
+    if (err == 0 || (len-off) < 130) {
+        error("%s", buf);
+        return;
     }
-    return 0;
+
+    ERR_load_crypto_strings();
```

Comment:
https://stackoverflow.com/questions/72165672/is-there-a-replacement-for-err-load-ssl-strings-since-it-is-deprecated

```There is no replacement. You don't need to call it anymore. Error strings are loaded automatically since OpenSSL 1.1.0.```

Reply:
I'm sorry, I am more into gnutls (but just as regular user) than openssl. Nice that these "global" call are now the default.